File: src/phylo/jax/pruning.py
import jax
import jax.numpy as jnp
from ._utils import token_to_likelihood


def upward_pass(alignment, tree, subMatrices, rootProb, maxChunkSize=128):
    """Felsenstein pruning: compute inside (U) vectors for all nodes.

    Args:
        alignment: (R, C) int32 tokens
        tree: Tree(parentIndex, distanceToParent)
        subMatrices: (*H, R, A, A) substitution probability matrices
        rootProb: (*H, A) root/equilibrium frequencies
        maxChunkSize: chunk columns to limit memory

    Returns:
        U: (*H, R, C, A) inside vectors per node (rescaled)
        logNormU: (*H, R, C) per-node subtree log-normalizers
        logLike: (*H, C) log-likelihoods
    """
    parentIndex = tree.parentIndex
    R, C = alignment.shape
    *H, _R, A, _A = subMatrices.shape
    assert _R == R and _A == A

    if C > maxChunkSize:
        results = []
        for i in range(0, C, maxChunkSize):
            chunk = alignment[:, i:i + maxChunkSize]
            U_c, lnU_c, ll_c = upward_pass(chunk, tree, subMatrices, rootProb, maxChunkSize)
            results.append((U_c, lnU_c, ll_c))
        U = jnp.concatenate([r[0] for r in results], axis=-2)
        logNormU = jnp.concatenate([r[1] for r in results], axis=-1)
        logLike = jnp.concatenate([r[2] for r in results], axis=-1)
        return U, logNormU, logLike

    # Initialize likelihood from tokens
    likelihood = token_to_likelihood(alignment, A)  # (R, C, A)
    if len(H) > 0:
        likelihood = jnp.broadcast_to(likelihood, (*H, R, C, A)).copy()
    else:
        likelihood = likelihood.copy()

    logNormU = jnp.zeros((*H, R, C))

    # Postorder branches: child R-1 down to 1
    child_indices = jnp.arange(R - 1, 0, -1, dtype=jnp.int32)
    parent_indices = jnp.flip(parentIndex[1:])
    sub_mats = jnp.flip(jnp.moveaxis(subMatrices, -3, 0)[1:, ...], axis=0)

    postorder = (child_indices, parent_indices, sub_mats)

    (likelihood, logNormU), _ = jax.lax.scan(
        _upward_step, (likelihood, logNormU), postorder
    )

    # Log-likelihood: log(sum_b pi_b * U_root_b) + logNormU[root]
    root_like = likelihood[..., 0, :, :]  # (*H, C, A)
    logLike = logNormU[..., 0, :] + jnp.log(
        jnp.einsum('...ca,...a->...c', root_like, rootProb)
    )

    # U is just the final likelihood array (rescaled inside vectors)
    U = likelihood
    return U, logNormU, logLike


def _upward_step(carry, branch):
    """Single step of the upward (pruning) scan."""
    likelihood, logNormU = carry
    child, parent, subMatrix = branch

    # Multiply child's contribution into parent:
    # parent_b *= sum_j M_bj * child_j
    child_contrib = jnp.einsum(
        '...ij,...cj->...ci', subMatrix, likelihood[..., child, :, :]
    )
    likelihood = likelihood.at[..., parent, :, :].multiply(child_contrib)

    # Rescale parent to prevent underflow
    maxLike = jnp.max(likelihood[..., parent, :, :], axis=-1)  # (*H, C)
    maxLike = jnp.maximum(maxLike, 1e-300)
    likelihood = likelihood.at[..., parent, :, :].divide(maxLike[..., None])
    log_rescale = jnp.log(maxLike)

    # Accumulate per-node log-normalizer:
    # parent's subtree norm += child's subtree norm + this rescaling
    logNormU = logNormU.at[..., parent, :].add(
        logNormU[..., child, :] + log_rescale
    )

    return (likelihood, logNormU), None

File: subby/wasm/src/pruning.rs
/// Upward pass (Felsenstein pruning).

use crate::token::token_to_likelihood;

/// Compute inside (U) vectors for all nodes.
///
/// Returns (U, logNormU, logLike):
///   U: (R*C*A) flat f64
///   logNormU: (R*C) flat f64
///   logLike: (C,) f64
pub fn upward_pass(
    alignment: &[i32],
    parent_index: &[i32],
    sub_matrices: &[f64],  // (R*A*A)
    root_prob: &[f64],     // (A,)
    r: usize,
    c: usize,
    a: usize,
) -> (Vec<f64>, Vec<f64>, Vec<f64>) {
    let mut u = token_to_likelihood(alignment, r, c, a);
    let mut log_norm_u = vec![0.0; r * c];

    // Postorder: children R-1 down to 1
    for n in (1..r).rev() {
        let p = parent_index[n] as usize;
        let m_base = n * a * a;

        for col in 0..c {
            // child_contrib[b] = sum_j M[b,j] * U[n,col,j]
            let child_base = (n * c + col) * a;
            let parent_base = (p * c + col) * a;

            for b in 0..a {
                let mut s = 0.0;
                for j in 0..a {
                    s += sub_matrices[m_base + b * a + j] * u[child_base + j];
                }
                u[parent_base + b] *= s;
            }

            // Rescale
            let mut max_val = u[parent_base];
            for b in 1..a {
                if u[parent_base + b] > max_val {
                    max_val = u[parent_base + b];
                }
            }
            if max_val < 1e-300 {
                max_val = 1e-300;
            }
            for b in 0..a {
                u[parent_base + b] /= max_val;
            }
            let log_rescale = max_val.ln();
            log_norm_u[p * c + col] += log_norm_u[n * c + col] + log_rescale;
        }
    }

    // Log-likelihood
    let mut log_like = vec![0.0; c];
    for col in 0..c {
        let mut s = 0.0;
        for aa in 0..a {
            s += root_prob[aa] * u[(0 * c + col) * a + aa];
        }
        log_like[col] = log_norm_u[0 * c + col] + s.ln();
    }

    (u, log_norm_u, log_like)
}
